pkg/scheduler/core/equivalence_cache.go

```diff
@@ -88,20 +85,28 @@ func (ec *EquivalenceCache) RunPredicate(
 	if err != nil {
 		return fit, reasons, err
 	}
-	// Skip update if NodeInfo is stale.
-	if cache != nil && cache.IsUpToDate(nodeInfo) {
-		ec.updateResult(pod.GetName(), nodeInfo.Node().GetName(), predicateKey, fit, reasons, equivClassInfo.hash)
+	if cache != nil {
+		ec.updateResult(pod.GetName(), predicateKey, fit, reasons, equivClassInfo.hash, cache, nodeInfo)
 	}
 	return fit, reasons, nil
 }
 
 // updateResult updates the cached result of a predicate.
 func (ec *EquivalenceCache) updateResult(
-	podName, nodeName, predicateKey string,
+	podName, predicateKey string,
 	fit bool,
 	reasons []algorithm.PredicateFailureReason,
 	equivalenceHash uint64,
+	cache schedulercache.Cache,
+	nodeInfo *schedulercache.NodeInfo,
 ) {
+	ec.mu.Lock()
+	defer ec.mu.Unlock()
+	// Skip update if NodeInfo is stale.
+	if !cache.IsUpToDate(nodeInfo) {
+		return
+	}
+	nodeName := nodeInfo.Node().GetName()
```

Comment:
Please add a check to ensure that nodeInfo.Node() is not nil before dereferencing it. It could be nil in tests.

Reply:
The caller is already checking that, but if this function is used from another place, the new caller may not check. So, it is better to check in the function.

Reply:
Yes, you are right. 
Fixed it, thanks!